docs/sphinx/examples/python/trotter_kernel_mode.py

```diff
@@ -0,0 +1,106 @@
+# ============================================================================ #
+# Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                   #
+# All rights reserved.                                                         #
+#                                                                              #
+# This source code and the accompanying materials are made available under     #
+# the terms of the Apache License 2.0 which accompanies this distribution.     #
+# ============================================================================ #
+
+import cudaq
+import time
+import matplotlib.pyplot as plt
+import numpy as np
+
+# Compute magnetization using Suzuki-Trotter approximation.
+# This example demonstrates usage of quantum states in kernel mode.
+#
+# Details
+# https://pubs.aip.org/aip/jmp/article-abstract/32/2/400/229229/General-theory-of-fractal-path-integrals-with
+#
+# Hamiltonian used
+# https://en.m.wikipedia.org/wiki/Quantum_Heisenberg_model
+
+
+# Alternating up/down spins
+@cudaq.kernel
+def getInitState(numSpins: int):
+    q = cudaq.qvector(numSpins)
+    for qId in range(0, numSpins, 2):
+        x(q[qId])
+
+
+# This performs a single-step Trotter on top of an initial state, e.g.,
+# result state of the previous Trotter step.
+@cudaq.kernel
+def trotter(state: cudaq.State, coefficients: list[complex],
+            words: list[cudaq.pauli_word], dt: float):
+    q = cudaq.qvector(state)
+    for i in range(len(coefficients)):
+        exp_pauli(coefficients[i].real * dt, q, words[i])
+
+
+def run_steps(steps: int, spins: int):
+    g = 1.0
+    Jx = 1.0
+    Jy = 1.0
+    Jz = g
+    dt = 0.05
+    n_steps = steps
+    n_spins = spins
+    omega = 2 * np.pi
+
+    def heisenbergModelHam(t: float) -> cudaq.SpinOperator:
+        tdOp = cudaq.SpinOperator(num_qubits=n_spins)
+        for i in range(0, n_spins - 1):
+            tdOp += (Jx * cudaq.spin.x(i) * cudaq.spin.x(i + 1))
+            tdOp += (Jy * cudaq.spin.x(i) * cudaq.spin.y(i + 1))
+            tdOp += (Jz * cudaq.spin.x(i) * cudaq.spin.z(i + 1))
+        for i in range(0, n_spins):
+            tdOp += (np.cos(omega * t) * cudaq.spin.x(i))
+        return tdOp
+
+    # Collect coefficients from a spin operator so we can pass them to a kernel
+    def termCoefficients(op: cudaq.SpinOperator) -> list[complex]:
+        result = []
+        ham.for_each_term(lambda term: result.append(term.get_coefficient()))
+        return result
+
+    # Collect Pauli words from a spin operator so we can pass them to a kernel
+    def termWords(op: cudaq.SpinOperator) -> list[str]:
+        result = []
+        ham.for_each_term(lambda term: result.append(term.to_string(False)))
+        return result
+
+    # Observe the average magnetization of all spins (<Z>)
+    average_magnetization = 0.0
+    for i in range(0, n_spins):
+        average_magnetization += ((1.0 / n_spins) * cudaq.spin.z(i))
+    average_magnetization -= 1.0
```

Comment:
Could you please check (e.g., printing out `average_magnetization`) that at the end of this code, the `spin_op` is `(1/N)*(Z0 + Z1 + ...)`?

The reason the C++ code has a -1 is because the [default constructor](https://github.com/NVIDIA/cuda-quantum/blob/2da801bc343a9cd5b62003e03cb24beb817b98c5/runtime/cudaq/spin/spin_op.cpp#L109) of `spin_op` is an identity term.

I'm not sure if it's also the case here.

Reply:
Will do, thanks!

Reply:
I had a couple of bugs - forgot to call the constructor (instead of initializing the `average_magnetization` to 0) and had a copy-paste bug in the hamiltonian. Fixed both, now the `average_magnetization` looks the same in c++ and python, and the results match as well.